### PR TITLE
[USMON-754] Improve http2 processing of headers

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?

- Introduce a simpler header processing mechanism. Rather than fully parsing 25 headers, we can concentrate on fully parsing only the initial 4 headers and introduce a simplified loop for the rest.
- Increase HTTP2_MAX_HEADERS_COUNT_FOR_FILTERING from 25 to 33 for processing more higher max headers in requests/responses. (for plain and tls)
- Parse HTTP2_MAX_PSEUDO_HEADERS_COUNT_FOR_FILTERING (4), bringing the total to 37. (for plain and tls)
- Increase HTTP2_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL from 15 to 19 for processing a higher number of headers in a single tail call. (plain)
- Decrease HTTP2_MAX_TAIL_CALLS_FOR_HEADERS_PARSER from 16 to 13, which allows us to use lass tail calls for the parsing part. (plain)
- Increase HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL from 10 to 15. (tls)
- Decrease HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER from 12 to 8, which allows us to use lass tail calls for the parsing part. (tls)

### Motivation

During the creation of the raw tests, we observed that headers have a specific (mandatory by the protocol) order, which we can leverage to improve our parsing.

> pseudo headers MUST appear in the header block before  regular header fields.

> Endpoints MUST NOT generate pseudo-header fields other than those defined in this document.

[ref](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.6)
[pseudo-headers](https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.1)

Thus, we can infer that headers like `:my-custom-header` will not appear, and only our “interesting headers” such as:
- `:scheme`
- `:method`
- `:authority`
- `:path`

Additionally, we are confident that we will encounter them before regular header fields. With this understanding, given that we have 4 such headers and the rest are different, we can optimize our parsing strategy.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 1]
```diff
@@ -13,10 +13,10 @@
 #define HTTP2_MAX_FRAMES_FOR_EOS_PARSER (HTTP2_MAX_FRAMES_FOR_EOS_PARSER_PER_TAIL_CALL * HTTP2_MAX_TAIL_CALLS_FOR_EOS_PARSER)
 
 // Represents the maximum number of frames we'll process in a single tail call in `handle_headers_frames` program.
-#define HTTP2_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL 15
+#define HTTP2_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL 23
 // Represents the maximum number of tail calls to process headers frames.
-// Currently we have up to 240 frames in a packet, thus 14 (15*16 = 240) tail calls is enough.
-#define HTTP2_MAX_TAIL_CALLS_FOR_HEADERS_PARSER 16
+// Currently we have up to 240 frames in a packet, thus 14 (11*23 = 253) tail calls is enough.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:18]:
```suggestion
// Currently we have up to 240 frames in a packet, thus 11 (11*23 = 253) tail calls is enough.
```

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 2]
```diff
@@ -27,14 +27,17 @@
 #define HTTP2_MAX_FRAMES_TO_FILTER  120
 
 // Represents the maximum number of frames we'll process in a single tail call in `uprobe__http2_tls_headers_parser` program.
-#define HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL 10
+#define HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL 15
 // Represents the maximum number of tail calls to process headers frames.
-// Currently we have up to 120 frames in a packet, thus 12 (12*10 = 120) tail calls is enough.
-#define HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER 12
+// Currently we have up to 120 frames in a packet, thus 12 (8*15 = 120) tail calls is enough.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:32]:
```suggestion
// Currently we have up to 120 frames in a packet, thus 15 (8*15 = 120) tail calls is enough.
```

File: pkg/network/ebpf/c/protocols/http2/decoding-defs.h [hunk 3]
```diff
@@ -27,14 +27,17 @@
...
+#define HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER 8
 #define HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER (HTTP2_TLS_MAX_FRAMES_FOR_HEADERS_PARSER_PER_TAIL_CALL * HTTP2_TLS_MAX_TAIL_CALLS_FOR_HEADERS_PARSER)
 
 // A limit of max headers which we process in the request/response.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-defs.h:36]: maybe elaborate here about the true meaning?

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 4]
```diff
@@ -63,6 +63,33 @@ static __always_inline bool tls_read_hpack_int(tls_dispatcher_arguments_t *info,
     return tls_read_hpack_int_with_given_current_char(info, current_char_as_number, max_number_for_bits, out);
 }
 
+// handle_non_pseudo_headers the case in which we are not in a pseudo header. we do not need to parse the header,
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:66]:
```suggestion
// Handles the case in which a header is not a pseudo header. We don't need to save it as interesting or modify our telemetry.
```

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 5]
```diff
@@ -59,6 +59,33 @@ static __always_inline bool read_hpack_int(struct __sk_buff *skb, skb_info_t *sk
     return read_hpack_int_with_given_current_char(skb, skb_info, current_char_as_number, max_number_for_bits, out);
 }
 
+// handle_non_pseudo_headers the case in which we are not in a pseudo header. we do not need to parse the header,
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:62]: take documentation from the TLS version

**Review by guyarb — COMMENTED** [review]
Overall looks great!

**Comment by amitslavin** [comment]
> // A limit of max headers which we process in the request/response.

Thanks :)

**Comment by amitslavin** [comment]
> Overall looks great!

THanks :)

**Review by guyarb — APPROVED** [review]

File: pkg/network/protocols/http2/raw_traffic.go [hunk 6]
```diff
@@ -14,17 +14,19 @@ import (
...
-// DynamicTableSize is the size of the dynamic table used in the HPACK encoder.
-const DynamicTableSize = 100
+// DynamicTableUpdateOptions is used to set the max dynamic table size.
+type DynamicTableUpdateOptions struct {
+	Size uint32
+}
 
 // NewHeadersFrameMessage creates a new HTTP2 data frame message with the given header fields.
-func NewHeadersFrameMessage(headerFields []hpack.HeaderField, setDynamicTableSize ...bool) ([]byte, error) {
+func NewHeadersFrameMessage(headerFields []hpack.HeaderField, setDynamicTableSize ...DynamicTableUpdateOptions) ([]byte, error) {
```
Comment (guyarb) [inline pkg/network/protocols/http2/raw_traffic.go:23]:
```suggestion
func NewHeadersFrameMessage(headerFields []hpack.HeaderField, dynamicTableOptions DynamicTableUpdateOptions) ([]byte, error) {
```

another option is:

```go
type headersFrameOptions struct {
  headers []hpack.HeaderField
  dynamicTableUpdate uint32
}
```

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 7]
```diff
@@ -147,30 +172,33 @@ static __always_inline __u8 filter_relevant_headers(struct __sk_buff *skb, skb_i
...
+        skb_info->data_off++;
+    #pragma unroll(HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS)
+        for (__u8 iter = 0; iter < HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS; ++iter) {
+            bpf_skb_load_bytes(skb, skb_info->data_off, &current_ch, sizeof(current_ch));
+            skb_info->data_off++;
+            if ((current_ch & 128) == 0) {
+                break;
+            }
+        }
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h:192]: worth maybe extracting it to a function, for a better readability

File: pkg/network/protocols/http2/raw_traffic.go (hunk 6 above)
Comment (amitslavin) [reply pkg/network/protocols/http2/raw_traffic.go:23]: Nice, liked it and changed it

File: pkg/network/ebpf/c/protocols/http2/decoding-tls.h [hunk 8]
```diff
@@ -154,6 +154,29 @@ static __always_inline bool tls_parse_field_literal(tls_dispatcher_arguments_t *
...
+    bpf_probe_read_user(&current_ch, sizeof(current_ch), info->buffer_ptr + info->data_off);
+    // If the top 3 bits are 001, then we have a dynamic table size update.
+    if ((current_ch & 224) == 32) {
+        info->data_off++;
+    #pragma unroll(HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS)
+        for (__u8 iter = 0; iter < HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS; ++iter) {
+            bpf_probe_read_user(&current_ch, sizeof(current_ch), info->buffer_ptr + info->data_off);
+            info->data_off++;
+            if ((current_ch & 128) == 0) {
+                break;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding-tls.h:174]:
```suggestion
                return;
```

File: pkg/network/ebpf/c/protocols/http2/decoding.h [hunk 9]
```diff
@@ -152,6 +152,29 @@ static __always_inline bool parse_field_literal(struct __sk_buff *skb, skb_info_
...
+    bpf_skb_load_bytes(skb, skb_info->data_off, &current_ch, sizeof(current_ch));
+    // If the top 3 bits are 001, then we have a dynamic table size update.
+    if ((current_ch & 224) == 32) {
+        skb_info->data_off++;
+    #pragma unroll(HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS)
+        for (__u8 iter = 0; iter < HTTP2_MAX_DYNAMIC_TABLE_UPDATE_ITERATIONS; ++iter) {
+            bpf_skb_load_bytes(skb, skb_info->data_off, &current_ch, sizeof(current_ch));
+            skb_info->data_off++;
+            if ((current_ch & 128) == 0) {
+                break;
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/http2/decoding.h]:
```suggestion
                return;
```

File: pkg/network/usm/usm_http2_monitor_test.go [hunk 10]
```diff
@@ -518,6 +519,9 @@ func (s *usmHTTP2Suite) TestRawTraffic() {
 					Method: usmhttp.MethodPost,
 				}: 1,
 			},
+			// Currently we don't have a way to test it as TLS version does not have the ability to run 2 tail calls
+			// for filtering frames (will be fixed in USMON-684)
+			skip: s.isTLS,
```
Comment (guyarb) [inline pkg/network/usm/usm_http2_monitor_test.go:524]: why?

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/usm/usm_http2_monitor_test.go (hunk 10 above)
Comment (amitslavin) [reply pkg/network/usm/usm_http2_monitor_test.go:524]: Mistake from the merge, thanks for the catch!

**Review by guyarb — APPROVED** [review]

**Pull request merged**